From 184c58a30e1ecabb087d858c3c7ef59e3fd14198 Mon Sep 17 00:00:00 2001
From: Yehuda Sadeh <yehuda@redhat.com>
Date: Mon, 20 Jul 2015 20:27:33 -0700
Subject: [PATCH 01/23] rgw: avoid using slashes for generated secret keys

Just use plain alphanumeric characterset.

Fixes: #7647

Signed-off-by: Yehuda Sadeh <yehuda@redhat.com>
(cherry picked from commit 3f04a6126fdbfb93304f798da3775c0eec9b7d44)
Signed-off-by: Dunrong Huang <dunrong.huang@eayun.com>

Conflicts:
	src/rgw/rgw_common.cc
	src/rgw/rgw_common.h
---
 src/rgw/rgw_common.cc | 20 ++++++++++++++++++++
 src/rgw/rgw_common.h  |  1 +
 src/rgw/rgw_user.cc   |  4 ++--
 3 files changed, 23 insertions(+), 2 deletions(-)

diff --git a/src/rgw/rgw_common.cc b/src/rgw/rgw_common.cc
index 5a1043f..9d4aa50 100644
--- a/src/rgw/rgw_common.cc
+++ b/src/rgw/rgw_common.cc
@@ -474,6 +474,26 @@ int gen_rand_alphanumeric(CephContext *cct, char *dest, int size) /* size should
   return 0;
 }
 
+static const char alphanum_plain_table[]="ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
+
+int gen_rand_alphanumeric_plain(CephContext *cct, char *dest, int size) /* size should be the required string size + 1 */
+{
+  int ret = get_random_bytes(dest, size);
+  if (ret < 0) {
+    lderr(cct) << "cannot get random bytes: " << cpp_strerror(-ret) << dendl;
+    return ret;
+  }
+
+  int i;
+  for (i=0; i<size - 1; i++) {
+    int pos = (unsigned)dest[i];
+    dest[i] = alphanum_plain_table[pos % (sizeof(alphanum_plain_table) - 1)];
+  }
+  dest[i] = '\0';
+
+  return 0;
+}
+
 int NameVal::parse()
 {
   int delim_pos = str.find('=');
diff --git a/src/rgw/rgw_common.h b/src/rgw/rgw_common.h
index 4d7a118..83799a2 100644
--- a/src/rgw/rgw_common.h
+++ b/src/rgw/rgw_common.h
@@ -178,6 +178,7 @@ enum {
 extern int gen_rand_base64(CephContext *cct, char *dest, int size);
 extern int gen_rand_alphanumeric(CephContext *cct, char *dest, int size);
 extern int gen_rand_alphanumeric_upper(CephContext *cct, char *dest, int size);
+extern int gen_rand_alphanumeric_plain(CephContext *cct, char *dest, int size);
 
 enum RGWIntentEvent {
   DEL_OBJ = 0,
diff --git a/src/rgw/rgw_user.cc b/src/rgw/rgw_user.cc
index 23575d8..3aae489 100644
--- a/src/rgw/rgw_user.cc
+++ b/src/rgw/rgw_user.cc
@@ -815,7 +815,7 @@ int RGWAccessKeyPool::generate_key(RGWUserAdminOpState& op_state, std::string *e
   } else if (gen_secret) {
     char secret_key_buf[SECRET_KEY_LEN + 1];
 
-    ret = gen_rand_base64(g_ceph_context, secret_key_buf, sizeof(secret_key_buf));
+    ret = gen_rand_alphanumeric_plain(g_ceph_context, secret_key_buf, sizeof(secret_key_buf));
     if (ret < 0) {
       set_err_msg(err_msg, "unable to generate secret key");
       return ret;
@@ -929,7 +929,7 @@ int RGWAccessKeyPool::modify_key(RGWUserAdminOpState& op_state, std::string *err
 
     int ret;
     int key_buf_size = sizeof(secret_key_buf);
-    ret  = gen_rand_base64(g_ceph_context, secret_key_buf, key_buf_size);
+    ret = gen_rand_alphanumeric_plain(g_ceph_context, secret_key_buf, key_buf_size);
     if (ret < 0) {
       set_err_msg(err_msg, "unable to generate secret key");
       return ret;
-- 
2.8.0.rc3

